apps/events/lib/events/gather.ex

```diff
@@ -4,19 +4,20 @@ defmodule Events.Gather do
   callbacks don't need to handle subsets of events.
   """
 
-  defstruct [:keys, :received, :callback]
+  defstruct [:keys, :callback, received: %{}]
 
   def new(keys, callback) when is_list(keys) and is_function(callback, 1) do
-    %__MODULE__{keys: Enum.sort(keys), received: %{}, callback: callback}
+    %__MODULE__{keys: MapSet.new(keys), callback: callback}
   end
 
   def update(%__MODULE__{keys: keys, received: received, callback: callback} = state, key, value) do
     received = Map.put(received, key, value)
 
-    if received |> Map.keys() |> Enum.sort() == keys do
+    if received |> Map.keys() |> MapSet.new() == keys do
```

Comment:
I'm not sure enough about MapSet to know if `==` works for equality. Should this be `MapSet.equal?/2`?

Also, is the only purpose of `keys` to keep track of if everything has been received yet? Would it be any faster if `update/3` here did a `MapSet.delete` whenever a new key came in, and then the conditional was to see if it was empty?

Reply:
In general no, `==` is not good enough because `MapSet` has two versions. However, it's only an issue if we're getting a MapSet from a previous version of Elixir which we aren't.

Interesting idea on checking the other way: I'll do a quick benchmark to find out.

Reply:
They're basically the same. `gather` is the current code, `reverse` is removing items from keys. The inputs are the number of keys we're checking for.

```
##### With input five #####
Name              ips        average  deviation         median         99th %
gather       579.97 K        1.72 μs  ±1500.58%        1.99 μs        2.99 μs
reverse      575.13 K        1.74 μs  ±1461.76%        1.99 μs        2.99 μs

Comparison:
gather       579.97 K
reverse      575.13 K - 1.01x slower +0.0145 μs

##### With input one #####
Name              ips        average  deviation         median         99th %
reverse        1.20 M      831.65 ns  ±3784.37%         990 ns         990 ns
gather         1.18 M      849.04 ns  ±3797.90%         990 ns         990 ns

Comparison:
reverse        1.20 M
gather         1.18 M - 1.02x slower +17.39 ns

##### With input ten #####
Name              ips        average  deviation         median         99th %
reverse      322.86 K        3.10 μs  ±1149.70%        2.99 μs        4.99 μs
gather       322.45 K        3.10 μs   ±915.33%        2.99 μs        5.99 μs

Comparison:
reverse      322.86 K
gather       322.45 K - 1.00x slower +0.00397 μs
```

Reply:
Ah, neat! Thanks for checking.